src/bp/core/repositories/events.ts

```diff
@@ -0,0 +1,64 @@
+import * as sdk from 'botpress/sdk'
+import { inject, injectable } from 'inversify'
+import _ from 'lodash'
+
+import Database from '../database'
+import { TYPES } from '../types'
+
+export interface EventRepository {
+  findEvents(fields: Partial<sdk.IO.StoredEvent>, searchParams: sdk.EventSearchParams)
+  deleteBeforeDate(date: Date): Promise<void>
+}
+
+export const DefaultSearchParams: sdk.EventSearchParams = {
+  sortOrder: [{ column: 'createdOn' }],
+  from: 0,
+  count: 10
+}
+
+const UNLIMITED_ELEMENTS = -1
+
+@injectable()
+export class KnexEventRepository implements EventRepository {
+  private readonly TABLE_NAME = 'events'
+
+  constructor(@inject(TYPES.Database) private database: Database) {}
+
+  async findEvents(
+    fields: Partial<sdk.IO.StoredEvent>,
+    params: sdk.EventSearchParams = DefaultSearchParams
+  ): Promise<sdk.IO.StoredEvent[]> {
+    if (!fields || !Object.keys(fields).length) {
+      throw new Error('At least one field is required')
+    }
+
+    const { sortOrder, count, from } = params
+
+    let query = this.database.knex(this.TABLE_NAME)
+    query = query.where(fields)
+
+    sortOrder &&
+      sortOrder.forEach(sort => {
+        query = query.orderBy(sort.column, sort.desc ? 'desc' : 'asc')
+      })
+
+    if (count !== UNLIMITED_ELEMENTS) {
+      query = query.limit(count)
+    }
+
+    return query.offset(from).then(rows =>
+      rows.map(storedEvent => ({
+        ...storedEvent,
+        event: this.database.knex.json.get(storedEvent.event)
+      }))
+    )
+  }
```

Comment:
* I know that @franklevasseur needs a groupBy filter. I'm not sure if he does it after query or not though ...

* We can't probably cover all use cases with a single function without re-inventing a orm..., does the user has access to raw query ?

Reply:
I've added a "generic" method which could cover most use case, it's always possible to add other queries in the repository, or to query the database directly for more advanced queries